veniq/dataset_collection/validation.py

```diff
@@ -0,0 +1,284 @@
+
+
+def find_extraction_opportunities(
+        method_ast: AST):
+    statements_semantic = extract_method_statements_semantic(method_ast)
+    extraction_opportunities = create_extraction_opportunities(statements_semantic)
+    filtered_extraction_opportunities = filter_extraction_opportunities(
+        extraction_opportunities, statements_semantic, method_ast
+    )
+    extraction_opportunities_groups = rank_extraction_opportunities(
+        statements_semantic, filtered_extraction_opportunities
+    )
+
+    return extraction_opportunities_groups
+
+
+@dataclass
+class RowResult:
+    output_filename: str
+    input_filename: str
+    start_line_SEMI: int
+    end_line_SEMI: int
+    start_line_dataset: int
+    end_line_dataset: int
+    percent_matched: float
+    class_name: str
+    method_name: str
+    error_string: str
+    ncss: int
+    matched: bool
+    failed_cases_in_SEMI_algorithm: bool
+    no_opportunity_chosen: bool
+    failed_cases_in_validation_examples: bool
+
+
+def fix_start_end_lines_for_opportunity(
+        extracted_lines_of_opportunity: List[int],
+        filepath: str) -> Tuple[int, int]:
+    """
+    Finds start and end lines for opportunity
+
+    :param filepath: filename where opportunity was found
+    :param extracted_lines_of_opportunity: list of lines for opportunity
+    :return: list of extracted lines for opportunity
+    """
+    start_line_opportunity = min(extracted_lines_of_opportunity)
+    end_line_opportunity = max(extracted_lines_of_opportunity)
+    text = read_text_with_autodetected_encoding(filepath).split('\n')
+    extraction = text[start_line_opportunity:end_line_opportunity]
+    open_brackets = 0
+    close_brackets = 0
+    for x in extraction:
+        close_brackets += x.count('}')
+        open_brackets += x.count('{')
+
+    if open_brackets < close_brackets:
+        diff = close_brackets - open_brackets
+        count = 1
+        for text_line in text[end_line_opportunity:]:
+            if diff < 1:
+                break
+            else:
+                if text_line.find('{') > -1:
+                    diff -= 1
+                    count += 1
+
+        start_line_opportunity += count - 1
+
+    elif open_brackets > close_brackets:
+        diff = open_brackets - close_brackets
+        count = 1
+        for text_line in text[end_line_opportunity:]:
+            if diff < 1:
+                break
+            else:
+                if text_line.find('}') > -1:
+                    diff -= 1
+                    count += 1
+
+        end_line_opportunity += count - 1
+
+    return start_line_opportunity, end_line_opportunity
+
+
+# flake8: noqa: C901
+def validate_row(dataset_dir: Path, row: pd.Series) \
+        -> List[RowResult]:
+    """
+    Validate row of dataset
+
+    :param dataset_dir: directory to dataset, path before the relative path in
+    output_filename
+    :param row: row of dataframe
```

Comment:
what dataframe?

Reply:
dataframe of the synth validation dataset

Reply:
can u please clarify in docstring

Reply:
added